README.md
# LoopBack REST Connector

LoopBack REST connector allows Node.js application to interact with HTTP REST APIs using a template driven approach.

> ### This project is modification of original [loopback-connector-rest](https://github.com/strongloop/loopback-connector-rest).

## Installation

npm install https://github.com/EdgeVerve/loopback-connector-rest/v1.0.0
> use latest version

## Options passed in before and after hooks of connector
Now OERest connector supports passing options (CallContext) to before and after hooks. It can be accessed as ctx.options.

## Supporting CRUD operations using template
Standard Rest Connecotr supports adding methods on Dao using templates however it does not support, templates for crud operations (settings.crud = true). To overcome this, OERest Connector now supports templates for methods on connector object (not on Dao), so that crud operations also can use template

## Rest Model Template 
Standard methods of rest models like create, update, updateAll, query, all etc. can now use templates only for modifying the request.

## Support of prototype methods 
In operations you can define prototype methods like updateAttributes, deleteById etc. These have advantage that this parameter will be pointed to instance, so you can use values like this.id in URL.
Example.
```
{
				"template": {
					"method": "PUT",
					"url": "http://localhost:4000/rest/bank/bank1/{model.modelName}/{this.id}",
					"body": "{!data:object}"
				},
				"functions": {
					"prototype.updateAttributes": ["data"]
				}
```


## Sample configuration

operations section is to templatize model custom methods, for templatizing inbuild crud methods, you can use templates section.

```
"oelocalrest": {
        "connector": "loopback-connector-rest",
        "debug": "false",
        "crud" : true,     
        "name": "oelocalrest",
        "updateAttributesByIdOnly" : false,
        "baseURL": "http://localhost:4000/api/",
        "options": {
            "headers": {
                "accept": "application/json",
                "content-type": "application/json"
            },
            "strictSSL": false
        },
        "templates": {
            "create": {
                "method": "POST",
                "uri": "http://localhost:4000/rest/bank/{!callContext.ctx.tenantId}/account",
                "body": "{!body:object}"
            },
            "query": {
                    "method": "GET",
                    "uri": "http://localhost:4000/rest/bank/{!callContext.ctx.tenantId}/account"
            }
        },
        "operations": [{
                "template": {
                    "method": "GET",
                    "url": "http://localhost:4000/rest/error1"
                },
                "functions": {
                    "error1": []
                }
        }
       ]
    }
```

## Documentation

For complete documentation, see [StrongLoop Documentation | REST Connector](http://docs.strongloop.com/display/LB/REST+connector).

CHANGES.md
2018-01-03, oe version 1.2.0
=========================
* Bug fixes

Previous
=========
* support for prototype methods in operations and special parameters in template this and model supported.




2016-03-18, Version 2.0.0
=========================

 * Derive the http source from the template (Raymond Feng)

 * Allow functions to specify param http source (Raymond Feng)


2016-03-15, Version 1.11.0
==========================

 * Upgrade deps (Raymond Feng)

 * Switch to jsonpath-plus (Raymond Feng)

 * This should fix issue #21 'Delete api not working' (Jouke Visser)

 * use an error object (Horia Radu)

 * handle http status code >= 4XX as errors (Horia Radu)


2016-02-19, Version 1.10.2
==========================

 * Remove sl-blip from dependencies (Miroslav Bajtoš)

 * Remove example dir (Simon Ho)

 * Refer to licenses with a link (Sam Roberts)

 * Use strongloop conventions for licensing (Sam Roberts)


2015-09-18, Version 1.10.1
==========================

 * Relax the test so that it matches both ave and avenue (Raymond Feng)


2015-07-10, Version 1.10.0
==========================

 * Use JSON.parse for default json values (Raymond Feng)


2015-05-27, Version 1.9.0
=========================

 * Add connector hooks (Raymond Feng)


2015-05-15, Version 1.8.1
=========================

 * fix issue #29 (MuSTa1nE)


2015-04-20, Version 1.8.0
=========================

 * Add bluebird as a dev dep (Raymond Feng)

 * Fix the assertions (Raymond Feng)

 * add promise support for rest-builder invoke method (Esco Obong)


2015-03-17, Version 1.7.0
=========================

 * Clean up the comments (Raymond Feng)

 * Remove commented out code (Raymond Feng)

 * Allow vars in keys for the template (Raymond Feng)

 * add lodash and merge defaults in order of precedence (Bryan Clark)


2015-02-23, Version 1.6.0
=========================

 * Upgrade deps (Raymond Feng)

 * Fix/workaround the test failure (Raymond Feng)

 * Fix bad CLA URL in CONTRIBUTING.md (Ryan Graham)


2014-11-27, Version 1.5.0
=========================

 * Add tests (Raymond Feng)

 * Allow options to be configured for request module (Raymond Feng)


2014-11-27, Version 1.4.1
=========================

 * Fix failing tests (Raymond Feng)


2014-10-06, Version 1.4.0
=========================

 * Bump version (Raymond Feng)

 * Add contribution guidelines (Ryan Graham)

 * Fix the samples to support LB 2.x (Raymond Feng)

 * Changed to propogate settings into rest-builder in order to support clientCert+Key (Shelby Sanders)


2014-09-26, Version 1.3.1
=========================

 * Update deps (Raymond Feng)

 * updateAttributes was double wrapping the callback in this.responseHandler causing it to be called twice which unexpected response parameter. (Arash Malekzadeh)


2014-08-29, Version 1.3.0
=========================

 * Bump version (Raymond Feng)

 * Upgrade to express 4.x (Raymond Feng)

 * Add connector type (Raymond Feng)

 * improve property existence test (Clark Wang)

 * Add form support to request builder (Clark Wang)

 * Allow to force json (Clark Wang)

 * Support specifying rest resource name (Lance Hudson)


2014-07-09, Version 1.2.1
=========================

 * Remove peer dep on loopback-datasource-juggler (Miroslav Bajtoš)

 * Update link to doc (Rand McKinney)

 * Reformat the code (Raymond Feng)

 * Add debug, derive http verb from the template (Raymond Feng)


2014-03-12, Version 1.1.4
=========================

 * Update dependencies and bump version (Raymond Feng)

 * Update to dual MIT/StrongLoop license (Raymond Feng)

 * Adds support for processors (Raymond Feng)

 * Replace old README with link to docs. (Rand McKinney)

 * Remove some of the dev deps for framework comparison examples (Raymond Feng)

 * Upgrade dependencies (Raymond Feng)


2013-12-06, Version 1.1.3
=========================

 * Bump version (Raymond Feng)

 * Update deps (Raymond Feng)

 * Update docs.json (Rand McKinney)

 * Fix the ref to Text type (Raymond Feng)

 * Remove blanket (Raymond Feng)

 * Bump version and update dependencies (Raymond Feng)

 * Fix package.json due to merging mistake (Raymond Feng)

 * Reset git dep (Raymond Feng)


2013-09-11, Version strongloopsuite-1.0.0-5
===========================================



2013-09-11, Version strongloopsuite-1.0.0-4
===========================================



2013-09-11, Version strongloopsuite-1.0.0-3
===========================================

 * Add keywords to package.json (Raymond Feng)


2013-09-10, Version strongloopsuite-1.0.0-2
===========================================

 * Capture more information to the error object (Raymond Feng)

 * Build an error ojbect only if err is not present (Raymond Feng)

 * Build an Error object (Raymond Feng)

 * Set up status code (Raymond Feng)

 * Finalize package.json for sls-1.0.0 (Raymond Feng)

 * Fix the test failure for not found (Raymond Feng)

 * Changed tag to strongloopsuite-1.0.0-2 (cgole)


2013-09-05, Version strongloopsuite-1.0.0-1
===========================================



2013-09-05, Version strongloopsuite-1.0.0-0
===========================================

 * Updated to use tagged version strongloopsuite-1.0.0-0 of dependencies (cgole)


2013-09-04, Version 1.2.0
=========================

 * First release!


test/rest-model.test.js
// Copyright IBM Corp. 2013,2016. All Rights Reserved.
// Node module: loopback-connector-rest
// This file is licensed under the MIT License.
// License text available at https://opensource.org/licenses/MIT

var assert = require('assert');

var ModelBuilder = require('loopback-datasource-juggler').ModelBuilder;
var modelBuilder = new ModelBuilder();

// simplier way to describe model
var User = modelBuilder.define('User', {
  name: String,
  bio: ModelBuilder.Text,
  approved: Boolean,
  joinedAt: Date,
  age: Number,
});

var RestResource = require('../lib/rest-model');

var rest = new RestResource('Users', 'http://localhost:3000');

describe('REST connector', function() {
  describe.skip('CRUD methods supported', function() {
    var server = null;
    before(function(done) {
      var app = require('./express-helper')();

      var count = 2;
      var users = [new User({ id: 1, name: 'Ray' }), new User({ id: 2, name: 'Joe' })];

      app.get('/Users', function(req, res, next) {
        res.setHeader('Content-Type', 'application/json');
        res.status(200).json(users);
      });

      app.post('/Users', function(req, res, next) {
        res.setHeader('Content-Type', 'application/json');
        var body = req.body;
        if (!body.id) {
          body.id = (++count);
        }
        res.setHeader('Location', req.protocol + '://' + req.headers['host'] + '/' + body.id);
        users.push(body);
        res.status(201).json(body);
      });

      app.put('/Users/:id', function(req, res, next) {
        for (var i = 0; i < users.length; i++) {
          var user = users[i];
          if (user.id == req.params.id) {
            res.setHeader('Content-Type', 'application/json');
            users[i] = req.body;
            res.status(200).end();
            return;
          }
        }
        res.status(404).end();
      });

      app.delete('/Users/:id', function(req, res, next) {
        for (var i = 0; i < users.length; i++) {
          var user = users[i];
          if (user.id == req.params.id) {
            res.setHeader('Content-Type', 'application/json');
            users.splice(i, 1);
            res.status(200).end();
            return;
          }
        }
        res.status(404).end();
      });

      app.get('/Users/:id', function(req, res, next) {
        for (var i = 0; i < users.length; i++) {
          var user = users[i];
          if (user.id == req.params.id) {
            res.setHeader('Content-Type', 'application/json');
            res.status(200).json(user);
            return;
          }
        }
        res.status(404).end();
      });

      server = app.listen(app.get('port'), function(err, data) {
        // console.log('Server listening on ', app.get('port'));
        done(err, data);
      });
    });

    after(function(done) {
      server && server.close(done);
    });

    it('should find two users', function(done) {
      rest.query(function(err, body, response) {
        assert.equal(200, response.statusCode);
        // console.log(body);
        assert.equal(2, body.length);
        done(err, body);
      });
    });

    it('should find the user with id 1', function(done) {
      rest.find(1, function(err, body, response) {
        assert.equal(200, response.statusCode);
        // console.log(err, response && response.statusCode);
        // console.log(body);
        assert.equal(1, body.id);
        assert.equal('Ray', body.name);
        done(err, body);
      });
    });

    it('should not find the user with id 100', function(done) {
      rest.find(100, function(err, body, response) {
        assert.equal(404, response.statusCode);
        // console.log(err, response && response.statusCode);
        // console.log(body);
        done(null, body);
      });
    });

    it('should update user 1', function(done) {
      rest.update(1, new User({ id: 1, name: 'Raymond' }), function(err, body, response) {
        assert.equal(200, response.statusCode);
        // console.log(err, response && response.statusCode);
        done(err, body);
      });
    });

    it('should delete user 1', function(done) {
      rest.delete(1, function(err, body, response) {
        assert.equal(200, response.statusCode);
        // console.log(err, response && response.statusCode);
        done(err, body);
      });
    });

    it('should create a new id named Mary', function(done) {
      rest.create(new User({ name: 'Mary' }), function(err, body, response) {
        assert.equal(201, response.statusCode);
        // console.log(response && response.statusCode);
        // console.log(response && response.headers['location']);
        // console.log(body);
        done(err, body);
      });
    });

    it('should list all users', function(done) {
      rest.query(function(err, body, response) {
        assert.equal(200, response.statusCode);
        // console.log(response && response.statusCode);
        // console.log(body);
        assert.equal(2, body.length);
        done(err, body);
      });
    });
  });
});


lib/rest-connector.js
// Copyright IBM Corp. 2013,2016. All Rights Reserved.
// Node module: loopback-connector-rest
// This file is licensed under the MIT License.
// License text available at https://opensource.org/licenses/MIT

/**
 *
 * ©2016-2017 EdgeVerve Systems Limited (a fully owned Infosys subsidiary),
 * Bangalore, India. All Rights Reserved.
 *
 */

var debug = require('debug')('loopback:connector:rest');
var RestResource = require('./rest-model');
var RequestBuilder = require('./rest-builder');
var request = require('request');
var _ = require('lodash');
var g = require('strong-globalize')();
var JsonTemplate = require('./template');

/**
 * Export the initialize method to loopback-datasource-juggler
 * @param {DataSource} dataSource The loopback data source instance
 * @param {function} [callback] The callback function
 */
exports.initialize = function initializeDataSource(dataSource, callback) {
  var settings = dataSource.settings || {};
  var baseURL = settings.baseURL || settings.restPath || 'http://localhost:3000/';

  var connector = new RestConnector(baseURL, settings);
  dataSource.connector = connector;
  dataSource.connector.dataSource = dataSource;
  connector.needOptionsArgument = true;

  var DataAccessObject = function() {};

    // Copy the methods from default DataAccessObject
  if (!settings.operations || settings.crud) {
    if (dataSource.constructor.DataAccessObject) {
      for (var i in dataSource.constructor.DataAccessObject) {
        DataAccessObject[i] = dataSource.constructor.DataAccessObject[i];
      }
            /* eslint-disable one-var */
      for (var i in dataSource.constructor.DataAccessObject.prototype) {
        DataAccessObject.prototype[i] = dataSource.constructor.DataAccessObject.prototype[i];
      }
            /* eslint-enable one-var */
    }
  }
  connector.DataAccessObject = DataAccessObject;

  if (Array.isArray(settings.operations)) {
    settings.operations.forEach(function(op) {
      if (!op.template) {
        throw new Error(g.f('The operation template is missing: %j', op));
      }
      var builder = RequestBuilder.compile(op.template, connector._request);
      builder.connector = connector;

            // Bind all the functions to the template
      var functions = op.functions;
      if (functions) {
        for (var f in functions) {
          if (debug.enabled) {
            debug('Mixing in method: %s %j', f, functions[f]);
          }
          var params = functions[f];
          var isPrototypeMethod = f.startsWith('prototype.');
          if (isPrototypeMethod) {
            f = f.substr(10);
          }
          var paramNames = [];
          var paramSources = [];
                    // The params can be ['x', 'y'] or [{name: 'x', source: 'header'},
                    // {name: 'y'}]
          for (var i = 0, n = params.length; i < n; i++) {
            if (typeof params[i] === 'string') {
              paramNames.push(params[i]);
              paramSources.push(null);
            } else if (typeof params[i] === 'object') {
              paramNames.push(params[i].name);
              paramSources.push(params[i].source);
            }
          }
          var fn = builder.operation(paramNames, isPrototypeMethod);
          dataSource[f] = fn;

          var path = '/' + f;
          fn.accepts = [];
          fn.shared = true;
          var args = builder.template.compile();
          paramNames.forEach(function(p, index) {
            var arg = args[p];
            var source = paramSources[index];
            if (!source) {
              source = arg.root;
              if (source === 'headers') {
                source = 'header';
              } else if (source === 'url') {
                source = 'path';
              }
            }
            if (source === 'path') {
                            /// Need to add path vars to the http url
              path += '/:' + p;
            }
            fn.accepts.push({
              arg: p,
              type: arg.type,
              required: arg.required,
              http: {
                source: source || 'query',
              },
            });
          });
          fn.returns = {
            arg: 'data',
            type: 'object',
            root: true,
          };
          fn.http = {
            verb: (op.template.method || 'GET').toLowerCase(),
            path: path,
          };
          fn.description = op.description || '';
          if (isPrototypeMethod) {
            DataAccessObject.prototype[f] = fn;
          }
          else {
            DataAccessObject[f] = fn;
          }
        }
      }
            // Inject the invoke function
      var invokeFn = function() {
        return builder.invoke.apply(builder, arguments);
      };
      var name = 'invoke';

      if (debug.enabled) {
        debug('Mixing in method: %s', name);
      }

      dataSource[name] = invokeFn;
      invokeFn.accepts = [
        {
          name: 'request',
          type: 'object',
        },
      ];
      invokeFn.shared = true;
            // dataSource.defineOperation(name, fn, fn);
      DataAccessObject[name] = invokeFn;
    });
  }

  settings.crudOperations = settings.crudOperations || {};

  Object.keys(settings.crudOperations).forEach(function(method) {
    var op = settings.crudOperations[method];
    var builder = RequestBuilder.compile(op, connector._request);
    builder.connector = connector;
    var params = op.args;
    var paramNames = [];
    var paramSources = [];
    for (var i = 0, n = params.length; i < n; i++) {
      if (typeof params[i] === 'string') {
        paramNames.push(params[i]);
        paramSources.push(null);
      } else if (typeof params[i] === 'object') {
        paramNames.push(params[i].name);
        paramSources.push(params[i].source);
      }
    }
    var fn = builder.connectorOperation(paramNames, op);
    dataSource[method] = fn;

        //var args = builder.template.compile();
    connector[method] = fn;
  });

  callback && process.nextTick(callback);
};

exports.RestConnector = RestConnector;

/**
 * The RestConnector constructor
 * @param {string} baseURL The base URL
 * @param {object} settings The settings
 * @constructor
 */
function RestConnector(baseURL, settings) {
  settings = settings || {};
  var options = settings.options || {};
  var defaults = settings.defaults || {};

  this._baseURL = baseURL;
  this._models = {};
  this._resources = {};

  debug('RestConnector:settings', settings);

  settings = _.omit(settings, ['options', 'defaults', 'connector', 'operations']);

    // Cascade the options in priority order for request module
    // options is the highest order then settings and defaults
  this._settings = _.merge(defaults, settings, options);

    // loopback-datasource-juggler uses connector.settings and
    // not connector._settings, other connectors like mongo,
    // postgres etc. also expose settings as settings, and not _settings
  this.settings = this._settings;
  var templates = this._settings.templates || {};
  var self = this;
  self.templates = {};
  Object.keys(templates).forEach(function(method) {
    var tempalte = new JsonTemplate(templates[method]);
    self.templates[method] = tempalte;
  });

    // map clientCert/clientKey to cert/key for backward compatibility
  if (this._settings.clientKey && this._settings.clientCert) {
    this._settings.key = this._settings.clientKey;
    this._settings.cert = this._settings.clientCert;
  }

  debug('RestConnector:this._settings', this._settings);
  this._request = request.defaults(this._settings);
}

/**
 * Hook for defining a model by the data source
 * @param {object} definition The model description
 */
RestConnector.prototype.define = function defineModel(definition) {
  var m = definition.model.modelName;
  this.installPostProcessor(definition);
  this._models[m] = definition;
  this._resources[m] = new RestResource(definition.settings.resourceName ||
        definition.model.pluralModelName, this._baseURL, this._request);
  var restModel = this._resources[m];
  restModel.connector = this;
  restModel.modelName = m;
};

RestConnector.prototype.getDefaultIdType = function() {
  return String;
};

/**
 * Install the post processor
 * @param {object} definition The model description
 */
RestConnector.prototype.installPostProcessor = function installPostProcessor(definition) {
  var dates = [];
  Object.keys(definition.properties).forEach(function(column) {
    if (definition.properties[column].type.name === 'Date') {
      dates.push(column);
    }
  });

  var postProcessor = function(model) {
    var max = dates.length;
    for (var i = 0; i < max; i++) {
      var column = dates[i];
      if (model[column]) {
        model[column] = new Date(model[column]);
      }
    }
    return model;
  };

  definition.postProcessor = postProcessor;
};

/**
 * Pre-process the request data
 * @param {*} data The request data
 * @returns {{}}
 */
RestConnector.prototype.preProcess = function preProcess(data) {
  var result = {};
  Object.keys(data).forEach(function(key) {
    if (data[key] !== null) {
      result[key] = data[key];
    }
  });
  return result;
};

/**
 * Post-process the response data
 * @param {string} model The model name
 * @param {*} data The response data
 * @param {boolean} many Is it an array
 */
RestConnector.prototype.postProcess = function postProcess(model, data, many) {
  var result = data;
  var postProcessor = this._models[model].postProcessor;
  if (postProcessor && data) {
    if (!many) {
      result = postProcessor(data);
    } else if (Array.isArray(data)) {
      result = [];
      var size = data.length;
      for (var i = 0; i < size; i++) {
        if (data[i]) {
          result[i] = postProcessor(data[i]);
        }
      }
    }
  }
  return result;
};

/**
 * Get a REST resource client for the given model
 * @param {string} model The model name
 * @returns {*}
 */
RestConnector.prototype.getResource = function getResourceUrl(model) {
  var resource = this._resources[model];
  if (!resource) {
    throw new Error(g.f('Resource for %s is not defined', model));
  }
  return resource;
};

/**
 * Create an instance of the model with the given data
 * @param {string} model The model name
 * @param {object} data The model instance data
 * @param {function} [callback] The callback function
 */
RestConnector.prototype.create = function create(model, data, options, callback) {
  this.getResource(model).create(data, options, function(err, body, response) {
    if (err) {
      callback && callback(err, body);
      return;
    }
    if (response.statusCode < 400) {
      body.statusCode = response.statusCode;
      callback && callback(null, body ? body.id : '', '' , body);
    } else {
      var err = g.f('Error response: %d %j', response.statusCode, body);
      callback && callback(err);
    }
  });
};

/**
 * Update or create an instance of the model
 * @param {string} model The model name
 * @param {object} data The model instance data
 * @param {function} [callback] The callback function
 * */
RestConnector.prototype.updateOrCreate = function(model, data, options, callback) {
  var self = this;
  this.exists(model, data.id, options, function(err, exists) {
    if (exists) {
      self.save(model, data, options, callback);
    } else {
      self.create(model, data, options, function(err, id) {
        data.id = id;
        callback(err, data);
      });
    }
  });
};

/**
 * A factory to build callback function for a response
 * @param {string} model The model name
 * @param {function} [callback] The callback function
 * @returns {function}
 */
RestConnector.prototype.responseHandler = function(model, callback, many) {
  var self = this;
  return function(err, body, response) {
    if (err) {
      callback && callback(err, body);
      return;
    }
    if (response.statusCode === 200) {
      if (callback) {
        var result = self.postProcess(model, body, many);
        callback(null, result);
      }
    } else {
      var err = g.f('Error response: %d %j', response.statusCode, body);
      callback && callback(err);
    }
  };
};

/**
 * Save an instance of a given model
 * @param {string} model The model name
 * @param {object} data The model instance data
 * @param {function} [callback] The callback function
 */
RestConnector.prototype.save = function save(model, data, options, callback) {
  this.getResource(model).update(data.id, data, options, this.responseHandler(model, callback));
};

RestConnector.prototype.update = function update(model, query, data, options, callback) {
    // PKG TODO Not sure why query.where should be checked
  if (data.id || query.where) {
    this.getResource(model).update(data.id, data, options, this.responseHandler(model, callback));
  } else {
    this.getResource(model).updateAll(query, data, options, this.responseHandler(model, callback));
  }
};

/**
 * Check the existence of a given model/id
 * @param {string} model The model name
 * @param {*} id The id value
 * @param {function} [callback] The callback function
 */
RestConnector.prototype.exists = function exists(model, id, options, callback) {
  this.getResource(model).find(id, options, function(err, body, response) {
    if (err) {
      callback && callback(err, body);
      return;
    }
    if (response.statusCode === 200) {
      callback && callback(null, true);
    } else if (response.statusCode === 404) {
      callback && callback(null, false);
    } else {
      var err = g.f('Error response: %s %j', response.statusCode, body);
      callback && callback(err);
    }
  });
};

/**
 * Find an instance of a given model/id
 * @param {string} model The model name
 * @param {*} id The id value
 * @param {function} [callback] The callback function
 */
RestConnector.prototype.find = function find(model, id, options, callback) {
  this.getResource(model).find(id, options, this.responseHandler(model, callback));
};

/**
 * Delete an instance for a given model/id
 * @param {string} model The model name
 * @param {*} id The id value
 * @param {function} [callback] The callback function
 */
RestConnector.prototype.destroy = function destroy(model, id, options, callback) {
  this.getResource(model).delete(id, options, this.responseHandler(model, callback));
};

/**
 * Query all instances for a given model based on the filter
 * @param {string} model The model name
 * @param {object} filter The filter object
 * @param {function} [callback] The callback function
 */
RestConnector.prototype.all = function all(model, filter, options, callback) {
  function getFields(data, arr) {
    _.forEach(data, function dataAccessGetKeysForEach(value, key) {
      if ((typeof key === 'string') && (key !== 'and' && key !== 'or')) {
        if (key.indexOf('.') > -1) {
          Array.prototype.splice.apply(arr, [0, 0].concat(key.split('.')));
        } else {
          arr.push({
            key: key,
            value: value,
          });
        }
      } else if (typeof value === 'object') {
        getFields(value, arr);
      }
    });
  }

  function isInstanceQuery(filter) { //pk name needs to be a parameter
    var pk = 'id';
    var whereConds = [];
    getFields(filter.where, whereConds);
    var _isDeleted = whereConds.find(function(cond) {
      return cond.key === '_isDeleted';
    });
    if (_isDeleted && _isDeleted.value === true) {
      return false;
    }
    var pkValue = whereConds.find(function(cond) {
      return cond.key === pk;
    });
    var allowedKeys = ['_scope', '_isDeleted'];
    if (pkValue !== undefined) {
      var modelAllowedKeys = [pk].concat(allowedKeys);
      var allowed = whereConds.reduce(function(result, cond) {
        if (!modelAllowedKeys.includes(cond.key)) {
          return false;
        } else {
          return result;
        }
      }, true);
      if (allowed) {
        return true;
      } else {
        return false;
      }
    } else {
      return false;
    }
  }

    // FIXME: [rfeng] Test the filter to decide if it's findById
  if (filter && filter.where && (!filter.order) &&
        filter.limit === 1 && filter.offset === 0) {
    if (isInstanceQuery(filter)) {
            // Map findById
      var whereConds = [];
      getFields(filter.where, whereConds);
      var id = whereConds.find(function(cond) {
        return cond.key === 'id';
      }).value;
      return this.find(model, id, options, function(err, result) {
        if (!err) {
          if (!result) {
            result = [];
          } else {
            result = [result];
          }
        }
        callback && callback(err, result);
      });
    }
  }
  this.getResource(model).all(filter, options, this.responseHandler(model, callback, true));
};

/**
 * Delete all instances for a given model
 * @param {string} model The model name
 * @param {function} [callback] The callback function
 */
RestConnector.prototype.destroyAll = function destroyAll(model, where, options, callback) {
  if (where && where.id) {
    this.getResource(model).delete(where.id, options, this.responseHandler(model, callback));
  } else {
    this.getResource(model).deleteAll(this.responseHandler(model, options, callback));
  }
};

/**
 * Count cannot not be supported efficiently.
 * @param {string} model The model name
 * @param {function} [callback] The callback function
 * @param {object} where The where object
 */
RestConnector.prototype.count = function count(model, callback, where) {
  throw new Error(g.f('Not supported'));
};

/**
 * Update attributes for a given model/id
 * @param {string} model The model name
 * @param {*} id The id value
 * @param {object} data The model instance data
 * @param {function} [callback] The callback function
 */
RestConnector.prototype.updateAttributes = function(model, id, data, options, callback) {
  data.id = id;
  this.getResource(model).update(id, options, this.responseHandler(model, callback));
};

/**
 * Get types associated with the connector
 * @returns {String[]} The types for the connector
 */
RestConnector.prototype.getTypes = function() {
  return ['rest'];
};
